File: src/metropolis.rs
//! Random-walk Metropolis on a declared energy, warm-up included, in one call (issue #1006).
//!
//! `sample.metropolis.random_walk` on the torch route makes one Python round
//! trip and one objective call per proposal. What compiles is a declared
//! family (`energy.rs`). The transition is `metropolis._RandomWalkKernel` at
//! unit temperature: `y = x + h s * z` with `z` standard normal and `s` the
//! metric's per-coordinate scale, accepted when a uniform on `[0, 1)` is
//! below `exp(U(x) - U(y))`. The warm-up is `chain._warm_up` operation for
//! operation, shared with HMC in `chain.rs`. The draws come from ChaCha8 seeded by the caller, so the
//! stream is this route's own and the torch route is matched in
//! distribution.

use numpy::{PyArray1, PyReadonlyArray1};
use pyo3::exceptions::PyValueError;
use pyo3::prelude::*;
use rand_chacha::ChaCha8Rng;
use rand_distr::{Distribution, StandardNormal, StandardUniform};

use crate::chain::{decide, Kernel, Walk, Warmup};
use crate::energy::Energy;
use crate::walk_class;

/// `metropolis._RandomWalkKernel`: where it is, its energy, and the buffers a step reuses.
pub struct RandomWalk {
    position: Vec<f64>,
    current: f64,
    proposal: Vec<f64>,
    scratch: Vec<f64>,
    scale: Vec<f64>,
}

impl RandomWalk {
    /// At `theta0` on `energy`, at unit scale.
    pub fn new(energy: &Energy<'_>, theta0: &[f64]) -> Self {
        let d = theta0.len();
        let mut scratch = vec![0.0; d];
        let current = energy.potential(theta0, &mut scratch);
        Self {
            position: theta0.to_vec(),
            current,
            proposal: vec![0.0; d],
            scratch,
            scale: vec![1.0; d],
        }
    }
}

impl Kernel for RandomWalk {
    #[inline]
    fn step(
        &mut self,
        energy: &Energy<'_>,
        rng: &mut ChaCha8Rng,
        step_size: f64,
    ) -> (bool, f64, f64) {
        for ((y, &x), &s) in self
            .proposal
            .iter_mut()
            .zip(&self.position)
            .zip(&self.scale)
        {
            let z: f64 = StandardNormal.sample(rng);
            *y = x + step_size * s * z;
        }
        let proposed = energy.potential(&self.proposal, &mut self.scratch);
        let uniform: f64 = StandardUniform.sample(rng);
        let (take, probability) = decide(self.current - proposed, uniform);
        let error = (proposed - self.current).abs();
        if take {
            std::mem::swap(&mut self.position, &mut self.proposal);
            self.current = proposed;
        }
        (take, probability, error)
    }

    fn position(&self) -> &[f64] {
        &self.position
    }

    fn set_scale(&mut self, scale: Vec<f64>) {
        self.scale = scale;
    }
}

/// `Walk::new` for the random walk from `theta0`.
pub fn walk(
    family: u8,
    parameters: Vec<f64>,
    theta0: &[f64],
    step_size: f64,
    seed: u64,
    warmup: Option<&Warmup>,
    powers: &[i32],
) -> Result<Walk<RandomWalk>, String> {
    Walk::new(
        family,
        parameters,
        theta0.len(),
        |energy| RandomWalk::new(energy, theta0),
        step_size,
        seed,
        warmup,
        powers,
    )
}

walk_class!(MetropolisWalk, RandomWalk, walk);

#[cfg(test)]
mod tests {
    use super::*;
    use crate::energy::GAUSSIAN;

    fn warmup(proposals: usize) -> Warmup {
        Warmup {
            proposals,
            target: 0.234,
            jitter: 0.0,
            gamma: 0.2,
            t0: 10.0,
            kappa: 0.75,
        }
    }

    #[test]
    fn the_chain_samples_the_diagonal_gaussian() {
        // Precisions 1 and 4: variances 1 and 0.25.
        let precision = [1.0, 4.0];
        let n = 200_000;
        let mut walk = walk(
            GAUSSIAN,
            precision.to_vec(),
            &[0.0, 0.0],
            1.2,
            1006,
            None,
            &[],
        )
        .unwrap();
        walk.advance(1_000, false, false);
        let result = walk.advance(n, true, false);
        for (coordinate, variance) in [(0, 1.0), (1, 0.25)] {
            let draws: Vec<f64> = result
                .draws
                .iter()
                .skip(coordinate)
                .step_by(2)
                .copied()
                .collect();
            let mean = draws.iter().sum::<f64>() / n as f64;
            let second = draws.iter().map(|x| x * x).sum::<f64>() / n as f64;
            assert!(mean.abs() < 0.03, "mean {mean}");
            assert!(
                (second - variance).abs() < 0.05 * variance.max(0.5),
                "E x^2 {second}"
            );
        }
    }

    #[test]
    fn the_warm_up_reaches_the_target_and_the_metric() {
        let precision = [1.0, 100.0];
        let result = walk(
            GAUSSIAN,
            precision.to_vec(),
            &[0.0, 0.0],
            0.1,
            7,
            Some(&warmup(4_000)),
            &[],
        )
        .unwrap();
        assert!((result.warmup_acceptance - 0.234).abs() < 0.05);
        // The mass diagonal estimates the precision to the window's sampling error.
        assert!((result.mass_diagonal[1] / result.mass_diagonal[0] / 100.0 - 1.0).abs() < 0.5);
    }

    #[test]
    fn blocks_continue_one_chain() {
        // Two blocks of 500 are one block of 1,000: the state and stream carry over.
        let make = || walk(GAUSSIAN, vec![1.0, 4.0], &[0.3, -0.2], 0.8, 11, None, &[]).unwrap();
        let whole = make().advance(1_000, true, false);
        let mut split = make();
        let (mut draws, a) = (
            split.advance(500, true, false),
            split.advance(500, true, false),
        );
        draws.draws.extend(a.draws);
        assert_eq!(draws.draws, whole.draws);
        assert_eq!(draws.accepted + a.accepted, whole.accepted);
    }

    #[test]
    fn the_filter_keeps_kalman_means_sums() {
        let mut walk = walk(
            GAUSSIAN,
            vec![1.0, 4.0],
            &[0.3, -0.2],
            0.8,
            5,
            None,
            &[1, 2],
        )
        .unwrap();
        let draws = walk.advance(300, true, true).draws;
        let f = &walk.filters[1];
        assert_eq!(f.n, 300);
        let ys: Vec<f64> = draws.iter().skip(1).step_by(2).map(|x| x * x).collect();
        assert_eq!(f.first[1], ys[0]);
        assert_eq!(f.last[1], ys[299]);
        let lagged: f64 = ys.windows(2).map(|w| w[0] * w[1]).sum();
        assert!((f.lagged[1] - lagged).abs() < 1e-12 * lagged.abs());
    }

    #[test]
    fn a_short_warm_up_and_a_bad_step_are_refused() {
        assert!(walk(GAUSSIAN, vec![1.0], &[0.0], 0.0, 0, None, &[]).is_err());
        assert!(walk(GAUSSIAN, vec![1.0], &[0.0], 0.1, 0, Some(&warmup(4)), &[]).is_err());
        assert!(walk(
            GAUSSIAN,
            vec![1.0, 2.0, 3.0],
            &[0.0, 0.0],
            0.1,
            0,
            None,
            &[]
        )
        .is_err());
    }
}

File: python/sal/sample/metropolis.py
"""Random-walk Metropolis over an :class:`~sal.opt.objective.Objective` (issue #1006).

The gradient-free chain: propose ``y = x + h z`` with ``z ~ N(0, T I)`` and
accept with probability ``min(1, exp((U(x) - U(y)) / T))``. The proposal is
symmetric, so the ratio is the energy difference alone (Metropolis et al.,
1953). One energy evaluation per proposal: the current point's energy is
carried from the step that reached it.

It is a :class:`~sal.sample.hmc.Kernel`, so
:func:`~sal.sample.hmc.run_chain` runs its warm-up, burn-in,
draws and operators, as it does for HMC and MALA. The warm-up is
:class:`~sal.sample.hmc.Adaptation` unchanged: a diagonal
metric is a per-coordinate proposal scale, and dual averaging drives the step
to a target acceptance, which belongs near :data:`RWM_TARGET_ACCEPTANCE`
(Roberts, Gelman & Gilks, 1997) and not at HMC's 0.65.

No derivative is taken, so the chain reads the objective through
:func:`~sal.opt.objective.energy_of` on arrays and draws from
a :class:`numpy.random.Generator` (issue #1011): an objective declaring a
NumPy :meth:`~sal.opt.objective.DeclaredEnergy.energy` is
evaluated without the tensor type, and any other through ``__call__``.

The compiled route (``oxisal.metropolis``, ``src/metropolis.rs``) runs the
whole chain, the warm-up included, on a declared family
(:mod:`sal.sample.declared`), from its own ChaCha8 stream.
:func:`replay` is the chain on randomness the caller supplies: the step
BlackJAX's ``rmh`` is pinned to draw for draw.
"""

from __future__ import annotations

import math
from collections.abc import Callable, Mapping
from dataclasses import dataclass
from typing import TYPE_CHECKING

import numpy as np

from sal import oxisal
from sal.backend import Backend, refuse_backend
from sal.emissions import ParameterDomainError
from sal.opt.objective import Objective, energy_of
from sal.sample import hmc
from sal.sample.accept import accept_ratio, acceptance_probability
from sal.sample.chain import (
    Adaptation,
    Transition,
    compiled_route,
    on_buffer,
    run_chain,
    run_compiled,
    start_point,
)
from sal.sample.declared import declared_energy

if TYPE_CHECKING:
    import torch

#: The asymptotically optimal acceptance for a random walk on a product
#: target as the dimension grows (Roberts, Gelman & Gilks, 1997).
RWM_TARGET_ACCEPTANCE = 0.234

#: Energy evaluations per proposal: the proposal's; the current point's is carried.
EVALUATIONS_PER_PROPOSAL = 1


def _decide(
    current: float, proposed: float, temperature: float, uniform: float
) -> tuple[bool, float, float]:
    """Accept or not, the acceptance probability, and the energy error."""
    log_ratio = (current - proposed) / temperature
    # ``math.exp`` overflows where a ratio is only "certainly accept"; a nan
    # energy stays nan, which the comparison refuses.
    ratio = math.exp(log_ratio) if not log_ratio > 700.0 else math.inf
    return (
        accept_ratio(ratio, uniform),
        acceptance_probability(ratio),
        abs(proposed - current),
    )


class _RandomWalkKernel:
    """One Gaussian proposal and its Metropolis test: :func:`random_walk`'s kernel.

    The generator is consumed as the normal increment, then the uniform. The
    step is on arrays: the loop's tensor is read through its own buffer and
    the proposal handed back on its buffer, so neither is copied. The current
    energy is kept beside the tensor it was taken at, and reused while the
    chain has not moved --- the identity of the tensor is the key, so a
    changed objective or a new position is evaluated afresh.
    """

    def __init__(self) -> None:
        self._at: tuple[Objective, torch.Tensor, float] | None = None

    def __call__(
        self,
        objective: Objective,
        position: torch.Tensor,
        temperature: float,
        generator: np.random.Generator,
        step_size: float,
    ) -> Transition:
        if (
            self._at is not None
            and self._at[0] is objective
            and self._at[1] is position
        ):
            current = self._at[2]
        else:
            current = energy_of(objective, position.numpy())
        increment = generator.standard_normal(position.shape[0])
        proposal = position.numpy() + step_size * math.sqrt(temperature) * increment
        try:
            proposed = energy_of(objective, proposal)
        except ParameterDomainError:
            proposed = math.inf
        uniform = float(generator.random())
        take, probability, error = _decide(current, proposed, temperature, uniform)
        if take:
            moved = on_buffer(proposal)
            self._at = (objective, moved, proposed)
            return Transition(moved, error, 1, probability)
        self._at = (objective, position, current)
        return Transition(position, error, 0, probability)


def random_walk(
    objective: Objective,
    rng: np.random.Generator,
    n_samples: int,
    *,
    step_size: float,
    theta0: np.ndarray | None = None,
    burn_in: int = 0,
    temperature: float = 1.0,
    adaptation: Adaptation | None = None,
    store_chain: bool = True,
    operators: Mapping[str, Callable[[torch.Tensor], torch.Tensor]] | None = None,
    backend: Backend = Backend.RUST,
) -> hmc.Chain:
    """Draw ``n_samples`` from ``exp(-objective / temperature)`` by random-walk Metropolis.

    Parameters
    ----------
    objective : Objective
        Read as an unnormalized negative log density; only its value is
        used, through :func:`~sal.opt.objective.energy_of`.
    rng : numpy.random.Generator
        The stream every increment and uniform is drawn from (issue #1011).
    n_samples : int
        Draws recorded after burn-in.
    step_size : float
        The proposal's standard deviation per coordinate, in the metric's
        coordinates. Required: its right value is the target's scale. With an
        ``adaptation`` it is the warm-up's starting point.
    theta0 : numpy.ndarray | None
        Starting point, copied; ``objective.initial()`` when omitted. An
        array, as the chain takes no derivative (issue #1059).
    burn_in, temperature, store_chain, operators
        As :func:`sal.sample.hmc.sample`.
    adaptation : Adaptation | None
        The two-window warm-up :func:`~sal.sample.hmc.sample`
        runs: dual averaging on the step, the variance for the metric. Its
        ``target_acceptance`` belongs near :data:`RWM_TARGET_ACCEPTANCE`.
    backend : Backend
        :data:`~sal.backend.Backend.RUST`, the default, runs
        the chain and its warm-up in ``oxisal.metropolis`` when the objective
        declares a family (:func:`~sal.sample.declared.declared_energy`)
        and the chain is at unit temperature in no tracked run; its
        ``operators`` observe the draws as :func:`~sal.sample.hmc.run_compiled`
        states, so a
        chain with ``store_chain=False`` holds that many draws at most. Its stream is ChaCha8 seeded by one draw from
        ``rng``, so it is pinned to the Python route in distribution.
        Any other chain, and :data:`~sal.backend.Backend.PYTHON`,
        run the NumPy kernel.

    Returns
    -------
    ~sal.sample.hmc.Chain
        The draws, acceptance, per-proposal energy error, energy evaluations
        spent (warm-up included) and what the warm-up settled on.

    Raises
    ------
    ValueError
        If ``step_size`` or ``temperature`` is not positive.
    """
    if not step_size > 0.0:
        msg = f"step_size must be positive, got {step_size}"
        raise ValueError(msg)
    refuse_backend("random_walk", backend, (Backend.PYTHON, Backend.RUST))
    declared = declared_energy(objective)
    if compiled_route(backend, temperature) and declared is not None:
        return run_compiled(
            oxisal.MetropolisWalk,
            declared,
            (),
            EVALUATIONS_PER_PROPOSAL,
            rng,
            n_samples,
            step_size=step_size,
            theta0=start_point(objective, theta0),
            burn_in=burn_in,
            adaptation=adaptation,
            store_chain=store_chain,
            operators=operators,
        )
    return run_chain(
        _RandomWalkKernel(),
        EVALUATIONS_PER_PROPOSAL,
        objective,
        rng,
        n_samples,
        step_size=step_size,
        theta0=theta0,
        burn_in=burn_in,
        temperature=temperature,
        adaptation=adaptation,
        store_chain=store_chain,
        operators=operators,
    )


@dataclass(frozen=True)
class Replayed:
    """A chain :func:`replay` ran: every position and whether each proposal was taken."""

    draws: np.ndarray
    accepted: np.ndarray


def replay(
    objective: Objective,
    theta0: np.ndarray,
    step_size: float | np.ndarray,
    increments: np.ndarray,
    uniforms: np.ndarray,
) -> Replayed:
    """The chain at unit temperature on the caller's increments ``(n, d)`` and uniforms ``(n,)``.

    :class:`_RandomWalkKernel`'s step with the draws supplied rather than
    taken from a generator, so an external implementation given the same
    randomness is compared draw for draw. ``step_size`` is a scalar or one
    scale per coordinate.
    """
    position = np.array(theta0, dtype=np.float64)
    scale = np.asarray(step_size, dtype=np.float64)
    current = energy_of(objective, position)
    draws = np.empty((len(uniforms), position.shape[0]))
    accepted = np.zeros(len(uniforms), dtype=bool)
    for index, (increment, uniform) in enumerate(
        zip(increments, uniforms, strict=True)
    ):
        proposal = position + scale * np.asarray(increment, dtype=np.float64)
        proposed = energy_of(objective, proposal)
        take, _, _ = _decide(current, proposed, 1.0, float(uniform))
        if take:
            position, current = proposal, proposed
        accepted[index] = take
        draws[index] = position
    return Replayed(draws, accepted)
